lucene/core/src/java/org/apache/lucene/codecs/lucene50/Lucene50PostingsReader.java

```diff
@@ -330,10 +331,15 @@ public PostingsEnum reset(IntBlockTermState termState, int flags) throws IOExcep
       skipped = false;
       return this;
     }
-    
+
+    // lazy loading of frequencies
     @Override
     public int freq() throws IOException {
-      return freq;
+      if (indexHasFreq && needsFreq && (isFreqDecoded == false)) {
```

Comment:
nit: this might be called for every doc while refillDocs is called more rarely. I'm wondering whether we could simplify this condition so that it would only be `if (isFreqDecoded == false) {`, ie. make sure that `isFreqDecoded is always `true` when the index doesn't have freqs or if freqs are not needed.